### PR TITLE
Export addons

### PR DESCRIPTION
**Pull request by EthanLowenthal**
# Goal
Enable easy loading of LabExT addons for additional export formats.

This PR should be merged after #214 (Addon Dialog)

# Approach
A wizard-style export dialog was added such that LabExT data can be exported to any format and event uploaded to a remote server. Each addon is respnsible for writing the data in any form it may choose, and may also overwrite the defualt "Export Settings" as well as "Exported Data Overview" steps. 

Be default the "Export Settings" steps propts the user for a directory path and makes the result available to the addon with `self.export_path.get()`. The default behaviour of the overview step is to display a treeview of the exported data. Both steps are visible in the screenshots below.

This defualt behaviour can be changed by overwriting the `build(self, frame)` and the `build_overview(self, frame)` methods, respectively.

# Implementation
This functionality makes use of the generic `Wizard` GUI element. The addons which are loaded for the manifest importer functionality must subclass `ExportStep` - which is in turn a step of the wizard. The addon class must provide an `_export(self, data)` method. Additionally, the addon class may override the `build` and `build_overview` methods. 

`_export(self, data)`  is called in a seperate thread in between the "Export Settings" and the "Exported Data Overview" steps.

# Compatibility
These changes did not break any tests. Two tests were added, as well as some documentation.

# Screenshots
#### 1st step - choosing which exporter to use and selecting data
Screenshot 2024-05-21 at 2 00 14 PM

#### 2nd step - exporter specific settings (by default, just an export folder path)
Screenshot 2024-05-21 at 2 00 23 PM

#### 3rd step - overview of exported data
Screenshot 2024-05-21 at 2 00 29 PM

### CODE REVIEW

File: LabExT/Exporter/ExportAPI.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+
+if TYPE_CHECKING:
+    from LabExT.ExperimentManager import ExperimentManager
+else:
+    ExperimentManager = None
+
+
+class ExportFormatAPI:
+    def __init__(self, experiment_manager: ExperimentManager):
```
Comment (DJAntivenom) [inline LabExT/Exporter/ExportAPI.py:23]: If you'd like you could add a doc-string here.

**Review by DJAntivenom — APPROVED** [review]
Very cool update :+1:

File: LabExT/Exporter/ExportStep.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
+
+
+
```
Comment (marcoep) [inline LabExT/Exporter/ExportStep.py:3]: whitespace

File: LabExT/Exporter/Formats/ExportHDF5.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                    group.create_dataset(k, data=v, dtype='f')
+                
+                metadata = measurement.copy()
+                metadata.pop("values")
+
+                write_metadata(file, metadata, '')
+                
+            file_names.append(oup_name)
+
+        self.wizard.logger.info('Exported %s files as .yaml: %s', len(file_names), file_names)
```
Comment (marcoep) [inline LabExT/Exporter/Formats/ExportHDF5.py:55]: typo

File: LabExT/Exporter/Formats/ExportJSON.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+This program is free software and comes with ABSOLUTELY NO WARRANTY; for details see LICENSE file.
+"""
+
+import json
+from os.path import join, exists
+from pathlib import Path
+
+from LabExT.Exporter.ExportStep import ExportFormatStep
+
+class ExportJSON(ExportFormatStep):
```
Comment (marcoep) [inline LabExT/Exporter/Formats/ExportJSON.py:14]: JSON exporter unnecessary as LabExT by default saves all measurements in JSON, can be removed.

File: LabExT/Tests/Wafer/Export_data_test.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        window._next_button.invoke()
+        self.pump_events()
+
+        file_path = join(dirname(dirname(__file__)), "example_measurement.csv")
+
+        self.assertTrue(exists(file_path))
+
+        with open(file_path) as f:
+            data = f.read()
+            self.assertTrue(data.startswith("# CSV exported measurement data from LabExT\n"))
```
Comment (marcoep) [inline LabExT/Tests/Wafer/Export_data_test.py:121]: Please add some more tests here, e.g. double check data content, other metadata?

File: LabExT/Tests/Wafer/Export_data_test.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        self.assertIsInstance(window.source_config_steps_insts["Hierarchical Data Format (.hdf5)"], ExportHDF5)
+
+    @pytest.mark.flaky(reruns=0)
+    def test_csv_format(self):
+        window = self.setup_window()
+        
+        window.current_step._select_all_button.invoke()
+        window.current_step.source_options_sel_var.set("Comma-Separated Values (.csv)")
+        self.pump_events()
+
```
Comment (marcoep) [inline LabExT/Tests/Wafer/Export_data_test.py:97]: here you might want to check that indeed one measurement is checked?

File: LabExT/Tests/Wafer/Export_data_test.py [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+    @pytest.mark.flaky(reruns=3)
+    def test_create_window_instantiate_steps(self):
+        window = self.setup_window()
+        self.assertIsInstance(window.source_config_steps_insts["JavaScript Object Notation (.json)"], ExportJSON)
+        self.assertIsInstance(window.source_config_steps_insts["Comma-Separated Values (.csv)"], ExportCSV)
+        self.assertIsInstance(window.source_config_steps_insts["Hierarchical Data Format (.hdf5)"], ExportHDF5)
+
+    @pytest.mark.flaky(reruns=0)
+    def test_csv_format(self):
```
Comment (marcoep) [inline LabExT/Tests/Wafer/Export_data_test.py:91]: Good job in finding out how to test-click through the GUI. I'd recommend adding some more asserts after every click to actually do testing.

**Review by marcoep — COMMENTED** [review]
Great PR, thank you!

File: LabExT/Exporter/ExportStep.py (hunk 2 above)
Comment (EthanLowenthal) [reply LabExT/Exporter/ExportStep.py]: Must have missed this

File: LabExT/Tests/Wafer/Export_data_test.py (hunk 5 above)
Comment (EthanLowenthal) [reply LabExT/Tests/Wafer/Export_data_test.py]: The test now checks the content of the data in the file against the measurement

File: LabExT/Tests/Wafer/Export_data_test.py (hunk 6 above)
Comment (EthanLowenthal) [reply LabExT/Tests/Wafer/Export_data_test.py]: I put a solution I think is good enough here, but its not foolproof.

File: LabExT/Exporter/Formats/ExportJSON.py (hunk 4 above)
Comment (EthanLowenthal) [reply LabExT/Exporter/Formats/ExportJSON.py]: This has been removed.

File: LabExT/Exporter/Formats/ExportHDF5.py (hunk 3 above)
Comment (EthanLowenthal) [reply LabExT/Exporter/Formats/ExportHDF5.py]: Fixed

**Pull request merged**